modules/dcache/src/main/java/org/dcache/services/ssh2/Ssh2Admin.java

```diff
@@ -316,6 +325,53 @@ public boolean authenticate(String userName, PublicKey key,
             }
             return successful;
         }
+
+        private boolean isValidHost(String line, ServerSession session) {
+            for (String linePart : Splitter.on(" ").trimResults().omitEmptyStrings().split(line)) {
+                if (linePart.startsWith("from=\"") && linePart.endsWith("\"")) {
+                    String from = linePart.substring(6, linePart.length()-1);
+                    Set<Outcome> outcomes = EnumSet.noneOf(Outcome.class);
+                    for (String pattern : Splitter.on(",").trimResults().omitEmptyStrings().split(from)) {
+                        outcomes.add(patternMatchesHost(pattern, session));
+                    }
+                    return outcomes.contains(Outcome.ALLOW) && !outcomes.contains(Outcome.DENY);
+                }
+            }
+            return true;
+        }
+
+        private Outcome patternMatchesHost(String pattern, ServerSession session) {
+            boolean patternIsNegated = pattern.startsWith("!");
+            Outcome matchOutcome = patternIsNegated ? Outcome.DENY : Outcome.ALLOW;
+            if (patternIsNegated) {
+                pattern = pattern.substring(1);
+            }
+            SocketAddress remote = session.getClientAddress();
+            if (remote instanceof InetSocketAddress) {
+                String remoteAddress = ((InetSocketAddress) remote).getAddress().getHostAddress();
```

Comment:
Since it is used in two places (and those lines are kinda long), you should probably factor out this cast; e.g.,

    SocketAddress rawRemote = session.getClientAddress();
    if (rawRemote instanceof InetSocketAddress) {
        InetSocketAddress remote = (InetSocketAddress) rawRemote;
        // ...
    }